Change search method in base class

oshtaier opened the pull request:
Removing search functionality from entity level and moving all logic on base level

Fixed Architecture:
```
nosetests tests/foreman/ui/test_architecture.py
..........................................F.
```
One failure due well-known issue with entity length (#2913)


Fixed GPG Key:
```
nosetests tests/foreman/ui/test_gpgkey.py
.SSSS.....E........S..S...EESE.ES.S..SE.
----------------------------------------------------------------------
Ran 40 tests in 5164.793s

FAILED (SKIP=10, errors=6)
```
Re-run after few more fixes:
```
nosetests tests/foreman/ui/test_gpgkey.py -m test_key_associate_15
.
----------------------------------------------------------------------
Ran 1 test in 150.347s

OK

nosetests tests/foreman/ui/test_gpgkey.py -m test_key_associate_16
.
----------------------------------------------------------------------
Ran 1 test in 183.642s

OK

nosetests tests/foreman/ui/test_gpgkey.py -m test_key_associate_17
.
----------------------------------------------------------------------
Ran 1 test in 197.358s

OK

nosetests tests/foreman/ui/test_gpgkey.py -m test_key_associate_19
.
----------------------------------------------------------------------
Ran 1 test in 199.676s

OK

nosetests tests/foreman/ui/test_gpgkey.py -m test_key_associate_20
.
----------------------------------------------------------------------
Ran 1 test in 204.137s

OK

nosetests tests/foreman/ui/test_gpgkey.py -m test_negative_update_1
.
----------------------------------------------------------------------
Ran 1 test in 424.711s

OK
```

oshtaier commented:
Ready for review and submit

abalakh commented on robottelo/ui/base.py:
```diff
@@ -163,14 +202,9 @@ def delete_entity(self, name, really, name_locator, del_locator,
         # verify that fact few times as sometimes 1 second is not enough for
         # element to be actually deleted from DB
+        self.is_entity_delete = True
         for _ in range(3):
```
I would suggest to wrap lines 206-214 with `try ... finally` block, to be 100% sure after exiting the `delete_entity()` variable will be set to `False`, so search function will know for sure it's not `delete_entity` called it.
I know it's 99% not an issue, i can barely imagine such case (negative delete test, which tries to delete and then searches for entity to verify it wasn't actually deleted?), so this comment is not a blocker.

oshtaier replied:
NACK on that comment as we will never experience any situation like that and will get variable into initial case in case of any failure
'negative delete test, which tries to delete and then searches for entity to verify it wasn't actually deleted?' is already covered here and will put variable into False state (`really` variable help us a lot)

Basically, don't want to introduce redundant block and make code less readable

sthirugn replied:
Actually +1 to @abalakh's question.  Is there a way to ensure that self.is_entity_delete is reset at all times inspite of any kind of failures on delete.

omaciel replied:
What both @abalakh and @sthirugn asked. Let's make sure to 'reset' values.

oshtaier replied:
Logically, we do not need it as per overall flow, there is a chance of 0.000001% that we will need to re-use search after we do delete and all necessary verification of entity existence. If we get a failure on any step - we will fail test case itself and proceed to the next (class variable value will be erased).

But, if you think that we need it - just tell and I will add try-finally block

oshtaier replied:
Okay, already 3 people:)

oshtaier replied:
Done

abalakh commented:
ACK
Great job :+1:

oshtaier commented:
@elyezer @omaciel @sthirugn @sghai @kbidarkar @lpramuk Please, review if you have time
Appreciate if you merge in case of second ACK, as it is very important for further progress

Sorry, if I push you too hard:) I just want to have proper results in time

sthirugn commented:
looking

sthirugn commented on robottelo/ui/base.py:
```diff
@@ -163,14 +202,9 @@ def delete_entity(self, name, really, name_locator, del_locator,
         # Make sure that element is really removed from UI. It is necessary to
         # verify that fact few times as sometimes 1 second is not enough for
         # element to be actually deleted from DB
+        self.is_entity_delete = True
```
Although I am not a fan of this workaround, its okay for now until we find out some other interesting way to do this.

oshtaier replied:
Absolutely agree with you!

omaciel commented on robottelo/ui/base.py:
```diff
@@ -64,31 +65,71 @@ def find_element(self, locator):
-        """Uses the search box to locate an element from a list of elements."""
-        search_key = search_key or 'name'
+    def _search_locator(self):
+        """Specify element name locator which should be use in search procedure
```
s/use/used

omaciel commented on robottelo/ui/base.py:
```diff
@@ -64,31 +65,71 @@ def find_element(self, locator):
+                     result_timeout=15):
+        """Initialize every optional parameter for search procedure which can
+        be overridden for any unique case"""
+        if self.is_entity_delete:
```
Could add a comment explaining the logic here? Why do we override the value for `button_timeout`? I'm sure there's a reason, it's just not clear to me from reading the code :)

sthirugn commented:
ACK pending question

omaciel commented on robottelo/ui/base.py:
```diff
@@ -64,31 +65,72 @@ def find_element(self, locator):
+        self.navigate_to_entity()
+
+        # Provide search criterions or use default ones
+        optional_parameters = self._search_init()
```
Do we need this `_search_init` function? Do you think we could move that logic here (it's not much really)?

oshtaier replied:
Yes, we need that method as we can like to override it in the child classes as we did for GPG key case or Product case:
```
+    def _search_init(self):
+        """Specify that Product entity has katello type"""
+        return super(Products, self)._search_init(is_katello=True)
```

sthirugn commented:
ACK. I will let @omaciel look and merge.

oshtaier commented:
Thank you @sthirugn

elyezer commented:
`_search_init` is being used to define if the entity is a katello one or not, what about creating a class attribute defaulting to False and then set to True for katello entities? Like:

```python
class BaseUI(object):
    is_katello = False
    # ...

class AKatelloEntity(BaseUI):
    is_katello = True
    # ...

class NotAKatelloEntity(BaseUI):
    # ...
```

abalakh commented:
@elyezer it's not only about katello, but also custom `search_key` or custom timeouts. So it's either 1 func or 4 class attributes.

oshtaier commented:
right

oshtaier commented:
and it is not so nice to add more things like that into class with name `Base`.

elyezer commented:
@abalakh @oshtaier can you please elaborate on why is not good to have class attributes for that?

oshtaier commented:
1) Method can have extra logic like we have here. Not sure, how I should implement same idea with attributes (when I specify 15 sec as value for generic case and 3 sec for specific case, but for the same entity)
2) Such attributes will be just perfect for class `Search`, but not `Base` as it is not only one focus for that class and it confuse people (that is a definitely a weak reason :))

Please, help, in case I do not see obvious things

elyezer commented:
@oshtaier @abalakh @omaciel @sthirugn all right, can we then rename the method to _search_parameters?

Agreed on the extra logic thing and using class attributes can cause issues since for any reason we change the Base value it will reflect on all classes that don't specify a value for it.

Thanks for the conversation.

oshtaier commented:
@elyezer Thank you for your inputs. I really like your way of thoughts, so I tried to implement proposed approach and now I know how to resolve that idea technically. Thanks to you
If you like to have it implemented at any moment - just let me know

Addressed your comment regarding name.

oshtaier commented:
Let me know if I need to change something else

omaciel commented:
@oshtaier got time for a hangout? Now :)

oshtaier commented:
sure

oshtaier commented:
Addressed all comments regarding class attributes idea.
Run one test to make sure that functionality works properly:
```
nosetests tests/foreman/ui/test_gpgkey.py -m test_key_associate_15
.
----------------------------------------------------------------------
Ran 1 test in 144.491s

OK
```
Run code in debug to make sure that delays really changed on '3' and '1' corrispondingly

Wanted to implement and merge today to not block 4 people tomorrow and wait till US time come

omaciel commented:
ACK

elyezer commented:
ACK
